### PR TITLE
Implement Questions Listing

### LINKED ISSUE #5: [Frontend] Implement Question Listing (Landing Page)
Display a list of questions on the app's main page. 
- Create components to render the questions in table form. (may be paginated)
- Fetch and display questions from database.

### LINKED ISSUE #7: [Backend] Set up API endpoint for questions listing page
* Set up the API endpoint to handle `GET` API requests for listing questions in the question index page
  * Return an array of question data as requested by the FE
  * Take care of edge cases/ invalid requests

**Sample JSON response:**
```typescript
export interface ApiResponse<D, M = {}> {
  code: null | number; // eg: 200, 400, 500...
  payload: {
		// metadata, eg: for get (index), how many entries were returned/ other relevant info
		meta: M;
		// the actual data
		data: D;
	};
  // [{type: enum- error/warning/success/info... (color of display msg), content: string}]
  messages: StatusMessage[];
}
```

### PR DESCRIPTION
**Pull request by chloeelim**
https://github.com/CS3219-AY2324S1/ay2324s1-course-assessment-g17/assets/79991214/ecf85f81-2873-40a0-b1c3-92f09899d3cb

Adds question listing frontend view, includes dark/light mode toggling, table sorting/ pagination and navbar.
[note: pagination in video demo above shows options 1 & 2 for demo purposes only- to show that the pagination works]

### CODE REVIEW

File: frontend/src/components/tables/DataTablePagination.tsx [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+        </Text>
+        <NumberInput
+          size="sm"
+          marginLeft={2}
+          marginRight={8}
+          width={28}
+          min={1}
+          max={getPageCount()}
+          onChange={(_str, num) => {
+            setPageIndex(Math.min(num - 1, 0));
```
Comment (seelengxd) [inline frontend/src/components/tables/DataTablePagination.tsx]:
```suggestion
            if (_str !== '') setPageIndex(Math.min(Math.max(num - 1, 0), getPageCount() - 1));
```

fixes issue where empty string causes NaN and all numbers set page to 1

(0:21 in the video)

**Review by seelengxd — COMMENTED** [review]
this looks very cool! :D

File: frontend/src/components/tables/DataTablePagination.tsx (hunk 1 above)
Comment (chloeelim) [reply frontend/src/components/tables/DataTablePagination.tsx]: Thanks for the catch! : D

File: frontend/src/components/tables/DataTablePagination.tsx [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+        </Text>
+        <NumberInput
+          size="sm"
+          marginLeft={2}
+          marginRight={8}
+          width={28}
+          min={1}
+          max={getPageCount()}
+          onChange={(_str, num) => {
+            if (_str !== '') setPageIndex(Math.max(num - 1, 0));
```
Comment (seelengxd) [inline frontend/src/components/tables/DataTablePagination.tsx]:
```suggestion
            if (_str !== '') setPageIndex(Math.min(Math.max(num - 1, 0), getPageCount() - 1));
```

still missing a page cap (test with 3)
(i edited the previous comment before so i think it didn't get added to the commit, my bad)

**Review by seelengxd — APPROVED** [review]
LGTM!

**Comment by sltsheryl** [comment]
the table looks really nice 🔥 

Screenshot 2023-09-10 at 9 16 04 PM

for some of the longer categories, the background is abit too small
(hmm but if its a desktop app and it's full screen, it's okay so are we handling responsiveness)

File: frontend/src/utils/questions.tsx [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+      header: 'Categories',
+      enableSorting: true,
+      cell: (categories) => (
+        <Stack direction="row">
+          {categories.getValue().map((category) => (
+            <Tag key={category}>{category}</Tag>
+          ))}
+        </Stack>
+      ),
+    }),
```
Comment (sltsheryl) [inline frontend/src/utils/questions.tsx]:
```js
columnHelper.accessor('categories', {
      header: 'Categories',
      enableSorting: true,
      cell: (categories) => (
        <Stack direction="row" spacing={4}>
          <Wrap>
            {categories.getValue().map((category) => (
              <WrapItem key={category}>
                <Tag>{category}</Tag>
              </WrapItem>
            ))}
          </Wrap>
        </Stack>
      ),
    }),
```
wraps the categories list to the next line if there are too many

Comment (sltsheryl) [reply frontend/src/utils/questions.tsx]: Screenshot 2023-09-10 at 10 01 11 PM

File: frontend/src/components/tables/DataTablePagination.tsx (hunk 2 above)
Comment (chloeelim) [reply frontend/src/components/tables/DataTablePagination.tsx]: Thanks for this!

File: frontend/src/utils/questions.tsx (hunk 3 above)
Comment (chloeelim) [reply frontend/src/utils/questions.tsx]: Thanks for this! Was trying to implement it but got stuck : " )

**Pull request merged**